cpp-ch/local-engine/Shuffle/SparkExchangeSink.h

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+#include <Processors/ISink.h>
+#include <Shuffle/SelectorBuilder.h>
+#include <Shuffle/ShuffleCommon.h>
+#include <jni.h>
+#include <Shuffle/PartitionWriter.h>
+
+namespace DB
+{
+    class QueryPipelineBuilder;
+}
+
+namespace local_engine
+{
+class CelebornClient;
+class PartitionWriter;
+
+class SparkExchangeSink : public DB::ISink
+{
+    friend class SparkExchangeManager;
+public:
+    SparkExchangeSink(const DB::Block& header, std::unique_ptr<SelectorBuilder> partitioner_,
+                      std::shared_ptr<PartitionWriter> partition_writer_,
+                      const std::vector<size_t>& output_columns_indicies_, bool sort_writer_)
+        : DB::ISink(header)
+          , partitioner(std::move(partitioner_))
+          , partition_writer(partition_writer_)
+          , output_columns_indicies(output_columns_indicies_)
+          , sort_writer(sort_writer_)
+    {
+        initOutputHeader(header);
+        partition_writer->initialize(&split_result, output_header);
+    }
+
+    String getName() const override
+    {
+        return "SparkExchangeSink";
+    }
+
+    SplitResult getSplitResultCopy() const
+    {
+        return split_result;
+    }
+
+    DB::Block getOutputHeaderCopy() const
+    {
+        return output_header.cloneEmpty();
+    }
```

Comment:
return const references, if we don't need to worry the content be modified in `SparkExechangeSink` during usage outside